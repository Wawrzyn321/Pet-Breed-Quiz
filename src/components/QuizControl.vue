<template>
  <div>
    <transition name="fade">
      <h3
        v-if="state === 'SHOWING_ANSWER'"
        class="link"
        @click="requestNextQuestion"
        id="request-next-question-button"
      >
        Next question >>
      </h3>
    </transition>

    <transition name="fade">
      <h3 v-if="state === 'FINISHED'" id="quiz-score-display">
        <p>Your score: {{ quizScore + " / " + questionsCount }}.</p>
      </h3>
    </transition>

    <p class="link" @click="finishQuiz" id="finish-quiz-button">
      Finish quiz
    </p>
  </div>
</template>

<script>
export default {
  methods: {
    requestNextQuestion() {
      this.$emit("next-question-requested");
    },
    finishQuiz() {
      this.$emit("finish-quiz-requested");
    }
  },
  props: {
    state: {
      required: true,
      type: String
    },
    quizScore: {
      required: false,
      type: Number,
      default: 0
    },
    questionsCount: {
      required: false,
      type: Number,
      default: 0
    }
  }
};
</script>

<style lang="scss" scoped>
h3,
p {
  text-align: center;
}

.link {
  cursor: pointer;
  text-decoration: underline;
}
</style>
