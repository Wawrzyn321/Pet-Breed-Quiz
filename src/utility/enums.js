const AppState = Object.freeze({
  Quiz: "QUIZ",
  Menu: "MENU"
});

const QuizState = Object.freeze({
  None: "NONE",
  ShowingQuestion: "SHOWING_QUESTION",
  ShowingAnswer: "SHOWING_ANSWER",
  Finished: "FINISHED"
});

const Categories = Object.freeze({
  Dogs: "DOGS",
  Cats: "CATS",
  Both: "BOTH",
  None: "NONE"
});

const Creatures = Object.freeze({
  Dog: "DOG",
  Cat: "CAT"
});

const AnswerState = Object.freeze({
  None: "NONE",
  Bad: "BAD",
  Good: "GOOD"
});

export { AppState, QuizState, Categories, Creatures, AnswerState };
