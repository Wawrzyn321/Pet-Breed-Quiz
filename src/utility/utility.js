const CONNECTION_TIMEOUT_MESSAGE = "Connection timeout";
const NETWORK_ERROR_MESSAGE = "NetworkError when attempting to fetch resource.";

function isConnectionError(error) {
  return (
    error.message === CONNECTION_TIMEOUT_MESSAGE ||
    error.message === NETWORK_ERROR_MESSAGE
  );
}

function getShuffledArray(sourceArray) {
  const array = [...sourceArray];
  for (let i = array.length - 1; i > 0; i--) {
    const j = getRandomNumber(i + 1);
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
}

function getRandomElements(array, count) {
  if (count > array.length) {
    throw new Error(
      "getRandomIndices: count must be less or equal the array length!"
    );
  }

  const shuffled = getShuffledArray(array);
  return shuffled.splice(0, count);
}

function getRandomNumber(maxExclusive) {
  return Math.floor(Math.random() * maxExclusive);
}

function fetchTimeout(url, options, timeout = 10000) {
  return Promise.race([
    fetch(url, options),
    new Promise((resolve, reject) =>
      setTimeout(() => reject(new Error(CONNECTION_TIMEOUT_MESSAGE)), timeout)
    )
  ]);
}

export {
  isConnectionError,
  getShuffledArray,
  getRandomElements,
  getRandomNumber,
  fetchTimeout
};
