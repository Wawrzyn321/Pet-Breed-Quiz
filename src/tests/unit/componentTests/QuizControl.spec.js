import QuizControl from "@/components/QuizControl";
import { shallowMount } from "@vue/test-utils";
import { QuizState } from "@/utility/enums";

describe("QuizControl", () => {
  describe("Rendering", () => {
    for (const state in QuizState) {
      it(`renders in ${state} state`, () => {
        const wrapper = shallowMount(QuizControl, {
          propsData: { state }
        });

        expect(wrapper.element).toMatchSnapshot();
      });
    }
  });

  describe("Events", () => {
    it("emits next questions request on clicking on next question button", () => {
      const wrapper = shallowMount(QuizControl, {
        propsData: { state: QuizState.ShowingAnswer }
      });

      wrapper.find("#request-next-question-button").trigger("click");

      expect(wrapper.emitted()).toEqual({ "next-question-requested": [[]] });
    });

    it("emits finish quiz request on clicking on finish quiz button", () => {
      const wrapper = shallowMount(QuizControl, {
        propsData: { state: QuizState.None }
      });

      wrapper.find("#finish-quiz-button").trigger("click");

      expect(wrapper.emitted()).toEqual({ "finish-quiz-requested": [[]] });
    });
  });

  describe("UI", () => {
    it("displays next question button on ShowingAnswer state", () => {
      const wrapper = shallowMount(QuizControl, {
        propsData: { state: QuizState.ShowingAnswer }
      });

      // eslint-disable-next-line prettier/prettier
      expect(wrapper.find("#request-next-question-button").isVisible()).toBe(true);
    });

    it("does not display next question button on state different from ShowingAnswer", () => {
      const wrapper = shallowMount(QuizControl, {
        propsData: { state: QuizState.ShowingQuestion }
      });

      expect(wrapper.find("#request-next-question-button").exists()).toBe(
        false
      );
    });

    it("displays quiz score when quiz is finished", () => {
      const wrapper = shallowMount(QuizControl, {
        propsData: { state: QuizState.Finished }
      });

      expect(wrapper.find("#quiz-score-display").isVisible()).toBe(true);
    });

    it("does not display quiz score when quiz is not finished", () => {
      const wrapper = shallowMount(QuizControl, {
        propsData: { state: QuizState.ShowingAnswer }
      });

      expect(wrapper.find("#quiz-score-display").exists()).toBe(false);
    });

    it("displays valid quiz score when quiz is finished", () => {
      const wrapper = shallowMount(QuizControl, {
        propsData: {
          state: QuizState.Finished,
          quizScore: 2,
          questionsCount: 10
        }
      });

      // eslint-disable-next-line prettier/prettier
      expect(wrapper.find("#quiz-score-display > p").text()).toBe("Your score: 2 / 10.");
    });
  });
});
