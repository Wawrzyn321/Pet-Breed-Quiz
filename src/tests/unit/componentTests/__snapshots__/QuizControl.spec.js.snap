// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizControl Rendering renders in Finished state 1`] = `
<div>
  <!---->
   
  <!---->
   
  <p
    class="link"
    id="finish-quiz-button"
  >
    
    Finish quiz
  
  </p>
</div>
`;

exports[`QuizControl Rendering renders in None state 1`] = `
<div>
  <!---->
   
  <!---->
   
  <p
    class="link"
    id="finish-quiz-button"
  >
    
    Finish quiz
  
  </p>
</div>
`;

exports[`QuizControl Rendering renders in ShowingAnswer state 1`] = `
<div>
  <!---->
   
  <!---->
   
  <p
    class="link"
    id="finish-quiz-button"
  >
    
    Finish quiz
  
  </p>
</div>
`;

exports[`QuizControl Rendering renders in ShowingQuestion state 1`] = `
<div>
  <!---->
   
  <!---->
   
  <p
    class="link"
    id="finish-quiz-button"
  >
    
    Finish quiz
  
  </p>
</div>
`;
